include/kitty/ashenhurst_decomposition.hpp

```diff
@@ -0,0 +1,201 @@
+/* kitty: C++ truth table library
+ * Copyright (C) 2017-2019  EPFL
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following
+ * conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*!
+  \file ashenhurst_decomposition.hpp
+  \brief Given a partition of function arguments, finds all of the ashenhurst decompositions.
+
+  \author Mahyar Emami (mahyar.emami@epfl.ch)
+*/
+
+#pragma once
+
+#include <cstdint>
+
+#include "constructors.hpp"
+#include "operations.hpp"
+#include <bitset>
+
+namespace kitty
+{
+
+namespace detail
+{
+
+/*! Composes a truth table
+  Given a function f represented as f(.), and a set of truth tables as arguments,
+  computes the composed truth table.
+  For example, if f(x1, x2) = 1001 and vars = {x1 = 1001, x2= 1010} the function
+  returns 1000 This function could be viewed as a general operator with arity 
+  vars.size() where the behavior of the operator is given by f.
+  \param f The outer function
+  \param vars The ordered set of input variables
+  \return The composed truth table with vars.size() variables
+*/
+template<class TTf, class TTv>
+auto compose_truth_table( const TTf& f, const std::vector<TTv> &vars )
+{
+  assert( vars.size() == f.num_vars() );
+  auto composed = vars[0].construct();
+  for ( uint64_t i = 0; i < composed.num_bits(); i++ )
+  {
+
+    uint64_t index = 0;
+    for ( uint64_t j = 0; j < vars.size(); j++ )
+    {
+      index += get_bit( vars[j], i ) << j;
+    }
+    auto bit = get_bit( f, index );
+    if ( bit == 1 )
+    {
+      set_bit( composed, i );
+    }
+    else
+    {
+      clear_bit( composed, i );
+    }
+  }
+  return composed;
+}
+/*! A helper function to enumerate missing indices
+  \param ys_index A list of already selected indices
+  \param max_index The maximum value for an index
+  \return Remaining indices
+*/
+std::vector<uint32_t>
+enumerate_zs_index( std::vector<uint32_t> ys_index, uint32_t max_index )
```

Comment:
`ys_index` should be passed `const&`

Reply:
Correct, missed that one! It is fixed now.